### PR TITLE
Allow checks to be configured using JSON file

### PR DESCRIPTION
**Pull request by stjj89**
Custom go_checker checks can now be configured using a JSON configuration file. The format of the configuration file will be specified in documentation that will be added in forthcoming commits.

### CODE REVIEW

File: go/tools/builders/generate_checker_main.go [hunk 1]
```diff
@@ -30,19 +32,43 @@ var codeTpl = `
 package main
 
 import (
-{{range $importPath := .}}
+{{- range $importPath := .ImportPaths}}
 	_ "{{$importPath}}"
 {{- end}}
 )
+
+// configs maps analysis names to configurations.
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/tools/builders/generate_checker_main.go]: Parsing the JSON here could be detrimental to performance if the configuration file has a lot of entries; code-generating a map and looking it up is definitely more efficient. The config-file-to-config-map transformation in generate_checker_main.go is relatively simple, and this template does not perform any complicated logic.

Is there any significant benefit to keeping the file around as a data dependency? Also, what about this template do you find hard to verify?

File: go/tools/builders/checker_main.go [hunk 2]
```diff
@@ -74,33 +77,72 @@ func run(args []string) error {
...
+			if len(config.applyTo) > 0 {
+				// This analysis applies exclusively to a set of files.
+				include = false
+				for pattern := range config.applyTo {
+					if matched, err := regexp.MatchString(pattern, filename); err == nil && matched {
+						include = true
+					}
+				}
+			}
+			for pattern := range config.whitelist {
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/tools/builders/checker_main.go]: When a file is "whitelisted" from a check, it means that the check does not apply to that file, therefore we discard any findings in that file. See the internal design doc for documentation on the configuration file. (Public docs coming soon.)

File: go/tools/builders/generate_checker_main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jayconrod) [reply go/tools/builders/generate_checker_main.go]: I don't have a strong opinion here; let's keep the template as-is.

I feel like templates in general aren't very readable due to the mix of code and data. Certainly it's strictly more expensive to parse JSON at run-time though, and all other things being equal, that should be the bigger concern.

We may want to revisit this in the future though if it becomes possible to use multiple configurations with the same checker binary.

[the remark this answers is not in the document]
Comment (stjj89) [reply go/tools/builders/generate_checker_main.go]: Ack.

File: tests/core/checks/custom/BUILD.bazel [hunk 3]
```diff
@@ -35,69 +37,136 @@ go_tool_library(
...
+        ":import_fmt",
     ],
+    extra_files = EXTRA_FILES,
+)
+
+bazel_test(
+    name = "custom_checks_no_errors",
+    build = BUILD_TMPL.format(config = ""),
+    workspace = WORKSPACE,
+    # check = NO_ERROR_CHECK,
```
[the remark this answers is not in the document]
Comment (stjj89) [reply tests/core/checks/custom/BUILD.bazel]: Done.

**Comment by jayconrod** [comment]
Looks like `custom_checks_custom_config` is failing. The message is below. (Are you able to access the test results?)

```
GoCompile: GoChecker: errors found during build-time code analysis:
/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/c375fa4a2a9ddd3b226bad434ee78a7b/execroot/bazel_test/external/io_bazel_rules_go/tests/core/checks/custom/foo_func.go:4:1: function must not be named Foo
```

Other failures are expected (Go 1.8 stuff from before).

**Comment by stjj89** [comment]
That error message is expected by custom_checks_custom_config. If the bazel_test itself failed, you should see a log message with the prefix "TEST FAILED". Can you paste the full bazel output of this test?

**Comment by jayconrod** [comment]
Here's the full output:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
-----------------------------------------------------------------------------
running in /tmp/tmp.aEm12W8QUY/isolate
TEST FAILED: received error message: 'custom/import_fmt.go:'
Loading: 
Loading: 0 packages loaded
Analyzing: 3 targets (18 packages loaded)
INFO: Analysed 3 targets (19 packages loaded).
INFO: Found 3 targets...
[0 / 4] [-----] BazelWorkspaceStatusAction stable-status.txt
INFO: From GoCompile external/io_bazel_rules_go/tests/core/checks/custom/linux_amd64_stripped/import_fmt~/importfmt.a:
GoChecker: errors found during build-time code analysis:
/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/c375fa4a2a9ddd3b226bad434ee78a7b/execroot/bazel_test/external/io_bazel_rules_go/tests/core/checks/custom/import_fmt.go:5:2: package fmt must not be imported


ERROR: /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/c375fa4a2a9ddd3b226bad434ee78a7b/external/io_bazel_rules_go/tests/core/checks/custom/BUILD.bazel:153:1: GoCompile external/io_bazel_rules_go/tests/core/checks/custom/linux_amd64_stripped/foo_func~/foofunc.a failed (Exit 1): compile failed: error executing command 
  (cd /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/c375fa4a2a9ddd3b226bad434ee78a7b/execroot/bazel_test && \
  exec env - \
  bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/linux_amd64_stripped/compile -go external/go_sdk/bin/go -root_file 'bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/ROOT' -goos linux -goarch amd64 '-cgo=1' -compiler_path /usr/bin -cc /usr/bin/gcc -c_flag -U_FORTIFY_SOURCE -c_flag -fstack-protector -c_flag -B/usr/bin -c_flag -B/usr/bin -c_flag -Wunused-but-set-parameter -c_flag -Wno-free-nonheap-object -c_flag -fno-omit-frame-pointer -c_flag -fno-canonical-system-headers -c_flag -Wno-builtin-macro-redefined -c_flag '-D__DATE__="redacted"' -c_flag '-D__TIMESTAMP__="redacted"' -c_flag '-D__TIME__="redacted"' -cxx_flag -U_FORTIFY_SOURCE -cxx_flag -fstack-protector -cxx_flag -B/usr/bin -cxx_flag -B/usr/bin -cxx_flag -Wunused-but-set-parameter -cxx_flag -Wno-free-nonheap-object -cxx_flag -fno-omit-frame-pointer -cxx_flag -fno-canonical-system-headers -cxx_flag -Wno-builtin-macro-redefined -cxx_flag '-D__DATE__="redacted"' -cxx_flag '-D__TIMESTAMP__="redacted"' -cxx_flag '-D__TIME__="redacted"' -cpp_flag -U_FORTIFY_SOURCE -cpp_flag -fstack-protector -cpp_flag -B/usr/bin -cpp_flag -B/usr/bin -cpp_flag -Wunused-but-set-parameter -cpp_flag -Wno-free-nonheap-object -cpp_flag -fno-omit-frame-pointer -cpp_flag -fno-canonical-system-headers -cpp_flag -Wno-builtin-macro-redefined -cpp_flag '-D__DATE__="redacted"' -cpp_flag '-D__TIMESTAMP__="redacted"' -cpp_flag '-D__TIME__="redacted"' -ld_flag '-fuse-ld=gold' -ld_flag -Wl,-no-as-needed -ld_flag -Wl,-z,relro,-z,now -ld_flag -B/usr/bin -ld_flag -B/usr/bin -ld_flag -pass-exit-codes -ld_flag -lstdc++ -ld_flag -lm -checker bazel-out/host/bin/linux_amd64_stripped/my_checker -package_list external/go_sdk/packages.txt -src external/io_bazel_rules_go/tests/core/checks/custom/foo_func.go -o 'bazel-out/k8-fastbuild/bin/external/io_bazel_rules_go/tests/core/checks/custom/linux_amd64_stripped/foo_func~/foofunc.a' -trimpath . -I . -- -p foofunc)
GoCompile: GoChecker: errors found during build-time code analysis:
/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/c375fa4a2a9ddd3b226bad434ee78a7b/execroot/bazel_test/external/io_bazel_rules_go/tests/core/checks/custom/foo_func.go:4:1: function must not be named Foo
INFO: Elapsed time: 1.749s, Critical Path: 0.38s
INFO: 5 processes, local.
FAILED: Build did NOT complete successfully
```

**Comment by stjj89** [comment]
It turns out the root cause of the issue was that the config file had an incorrect entry for a check named `"import_unsafe"`, which does not exist. Changing this to `"import_fmt_check"` should have fixed the issue. However, I ran into problems getting the test to pass locally even after this making this change.

I think this has to do with Bazel (or `bazel_test`?) caching behavior---a mere change to the config file does not seem sufficient to cause the checker in the `bazel_test` environment to be re-generated. I spent some time trying to reverse-engineer this caching logic, but I couldn't figure it out. Do you know what might be causing this issue? I definitely want the checker binary to be re-generated each time a change is made to the config file.

**Comment by jayconrod** [comment]
It looks like the issue here is that the file names in `extra_files` affect the generated script (it needs to know which files to copy), but nothing cares about the contents of those files.

I've pushed #1524, which adds `extra_files` to the `bazel_test` on master. It puts the files into the `runfiles` for `_bazel_test_script`. The `sh_test` that depends on that should pick up those files and re-execute the test when they change. You'll need to make a similar change on the `check` branch.

```
    return struct(
        files = depset([script_file]),
        runfiles = ctx.runfiles(
            [workspace_file, build_file] + ctx.files.extra_files,
            collect_data = True,
        ),
    )
```

**Comment by stjj89** [comment]
Done. Thanks for looking into this and submitting that change on master.

**Pull request merged**